Feature/full harvest using mutatiedienst

bdevloed opened the pull request:
CLBV-1041

Validation logic is kept inside `is_valid_association_data(item)`

If an association matches any of these criteria, the data will be skipped and not included in the verenigingingen app:
- 404 response, with the response body specifying it has been removed.
- 502 Bad Gateway error AND feature flag `FEATURE_SKIP_UNEXPECTED_RESPONSE_FROM_SOURCE: true`
  - Due to data validation issues
- No locations
- No locations in Flanders or Brussels - note that this does mean that as long as there is at least one location in Flanders or Brussels capital region, it's considered valid. And locations in the verenigingen-app could therefore be from outside of these regions.
- Has been marked as double

Note on the postal code check:
Check is done based on a list of postal codes, not a simpler range check. There are subtle differences such as special postal codes. e.g.  `1031 | Christelijke Sociale Organisaties` which is not in the list, bu

- [x] deleted
- [x] Double/duplicate: skipping for now - to be confirmed whether this is desired.
- [x] no postal code
- [x] Wallonia
- [x] Special postal code
- [x] Outside of BE

use transformation test script:
```
docker run --rm -v "$PWD:/app" -w /app semtech/mu-python-template:2.0.0-beta.2 python tests/test_transform_data.py
```

bdevloed committed:
check token expiration

bdevloed committed:
Not all http error responses contain a status attribute

bdevloed committed:
use mutatiedienst instead of postal code search

bdevloed committed:
fix imports

bdevloed committed:
skip duplicate associations (for now)

bdevloed committed:
test script

bdevloed committed:
feature flag to skip unexpected resources and backoff on ratelimit errors

bdevloed committed:
log statement

bdevloed committed:
get access token at request time

bdevloed committed:
Grouped and extended data validation checks

bdevloed committed:
include isUitgeschrevenUitPubliekeDatastroom associations

cecemel commented on lblod/detail_fetcher.py:
```diff
@@ -55,11 +54,26 @@ def fetch_detail_url(access_token, v_code, task):
+                    continue
+                else:
+                    logger.error(f"Rate limit (429) persisted after {retry_attempts} attempts for vCode {v_code}")
+                    break
```
I think i have the same remark for alle supressed failures
https://github.com/lblod/harvesting-verenigingen-scraper-service/pull/24/files#diff-f3cc572e9b4daa5354583639717cb61eb2b87c353614715347d04715b4bab36aR48-R53

Unless we accidentally break stuff elsewhere, we run at risk of  skipping these resources no?

cecemel commented on lblod/helpers.py:
```diff
@@ -3,13 +3,43 @@
+    return None
+
 def get_access_token():
+    global _cached_authentication
```
I am not sure why it has been added. 
I am not sure how this is going to work in multi-threaded environment...
I am just going to accept it now, it's too unknown territory for  me.

bdevloed replied:
On the checks:
- We should be careful to continue when an unexpected error is returned. The harvesting code should e.g. not just continue in case the API starts returning 404s because it's down for maintenance or some error.

This could lead to e.g. all associations being deleted downstream.

bdevloed replied:
I don't see an issue with the cached token, or the use of global as a keyword.

Also, we can re-use the token over multiple threads but also could use multiple tokens over multiple threads. So even if there's a logical error in that part, we're still ok I think :)

Caching itself is needed to avoid re-requesting a token many thousands of times during every harvest

cecemel replied:
I see hat the idea could be a win. I am rather wondering whether we had a problem with that beforehand?
Furthermore, that multiple threads could modify the one variable, don't we have potentially issues with race conditions?  Again I am wary, because i don't know enough about threaded programming. (Also I am more used to the multiprocessing packages, vs threads) 
Anyway, we can start with that.

bdevloed replied:
We didn't have a problem yet. A token was fetched and used for the scraping. No checks on expiry. 
The expiry is currently 1h, but could change. Also more and more associations are being created. We still had margin, but I didn't want to wait until we get the problem to fix it.

cecemel committed:
Attempt at simplification - remove unused task variable

cecemel committed:
Attempt a simplification ->

 - Uniform sleep time for all errors
 - Extract the sleep time calculation to another function
 - split fetch_detail_url in two functions
    - one busy with retry
    - one busy with attempting to fetch the detail

cecemel commented on lblod/detail_fetcher.py:
```diff
@@ -5,16 +5,15 @@
 
 
-def fetch_detail_url(access_token, v_code, task):
+def fetch_detail_url(v_code, task):
```
I made an attempt here at simplifying a bit the flow of control.
https://github.com/lblod/harvesting-verenigingen-scraper-service/pull/25

cecemel commented on lblod/transform_data.py:
```diff
@@ -217,3 +248,31 @@ def create_representative(representative_data, v_code):
+
+
+
+@lru_cache(maxsize=1)
```
No idea what this is.

cecemel replied:
After asking an LLM I understand.
Okay, there are about 1000 postal codes in Belgium? 
I don't think all the engineering is needed to deal with a list of 1000 entries in memory. 
Why has it been added?

cecemel replied:
( I would leave it as is, but I am curious about the motivation here)

bdevloed replied:
without this: a file is read, parsed for each and every association which is being transformed. 30k times per harvest
with this - use a built in mechanism to do this only once. It's just a simple decorator. imho not a lot of engineering.

cecemel replied:
Loading at starup in a global variabele wasn't an option?

bdevloed replied:
That's an alternative. I found this easy - but that's a matter of taste I guess?

cecemel reviewed (changes requested):
I left some  questions. I suggested some changes.

But testing I got:  
```
harvest_scraper-1  | 2025-11-26T14:55:10.520259186Z Unexpected http error: {'type': 'urn:associationregistry.admin.api:validation', 'title': 'Er heeft zich een fout voorgedaan!', 'detail': 'Source: De gevraagde vereniging werd niet gevonden.', 'status': 404, 'instance': '27398d1e-06b5-4651-9a18-44e11c3e2c45'}
```

This fails the full pipeline. Maybe we should do less strict checking on `404`.
Maybe this can be kept in a separate report or so
I remove the strict checks and it worked then

bdevloed committed:
do not reuse correlation_id

bdevloed committed:
Merge pull request #25 from lblod/feature/full-harvest-using-mutatiedienst-corrections

Feature/full harvest using mutatiedienst corrections

bdevloed commented:
Discussed merge / approval with @cecemel
